sdk/core/azure-core/src/main/java/com/azure/core/util/Configuration.java

```diff
@@ -214,19 +184,45 @@
      */
     @SuppressWarnings("StaticInitializerReferencesSubClass")
     public static final Configuration NONE = new NoopConfiguration();
+    private static final String[] EMPTY_ARRAY = new String[0];
 
-    private final ConcurrentMap<String, Optional<String>> configurations;
+    private final EnvironmentConfiguration environmentConfiguration;
+    private final Map<String, String> configurations;
+    private final String path;
+    private final Configuration sharedConfiguration;
+    private final ClientLogger logger;
+    private final boolean isEmpty;
 
     /**
      * Constructs a configuration containing the known Azure properties constants.
+     *
+     * @deprecated Use {@link ConfigurationBuilder} and {@link ConfigurationSource}  that allow to
+     * provide all properties before creating configuration and keep it immutable.
      */
+    @Deprecated
     public Configuration() {
-        this.configurations = new ConcurrentHashMap<>();
-        loadBaseConfiguration(this);
+        this(Collections.emptyMap(), EnvironmentConfiguration.getGlobalConfiguration(), null, null);
+    }
+
+    /**
+     * Constructs a configuration containing the known Azure properties constants. Use {@link ConfigurationBuilder} to create instance of {@link Configuration}.
+     *
+     * @param configurations map of all properties.
+     * @param environmentConfiguration instance of {@link EnvironmentConfiguration} to mock environment for testing.
+     * @param path Absolute path of current configuration section for logging and diagnostics purposes.
+     * @param sharedConfiguration Instance of shared {@link Configuration} section to retrieve shared properties.
+     */
+    Configuration(Map<String, String> configurations, EnvironmentConfiguration environmentConfiguration, String path, Configuration sharedConfiguration) {
+        this.configurations = Collections.unmodifiableMap(Objects.requireNonNull(configurations, "'configurations' cannot be null"));
```

Comment:
should we copy the map ? the unmodifiableMap doesn't prevent mutation if caller of ctor retains reference to the argument.

Reply:
refactored it to the point where we read properties in Configuration ctor and can guarantee underlying map won't be modified